python/rainbow/simulators/prox_soft_bodies/gauss_seidel_parallel.py

```diff
@@ -0,0 +1,381 @@
+import numpy as np
+import numba as nb
+import networkx as nx
+from collections import defaultdict
+from rainbow.util.timer import Timer
+
+
+def _build_contact_graph(J):
+    """ Create the contact graph based on the contact jacobi matrix.
+        The vertices of the graph are the contact points, and the edges of the graph are the value of the contact jacobi matrix is not zero between two contact points.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        graph: The contact graph.
+    """
+    G = nx.Graph()
+    K = J.shape[0] // 4
+    G.add_nodes_from(np.arange(K))
+    sources, targets = _get_edges(J)
+    for s, t in zip(sources, targets):
+        G.add_edge(s, t) 
+    
+    return G
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def _get_edges(J):
+    """ Get the edges of the contact graph.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        (ArrayLike, ArrayLike): The source and target of the edges.
+    """
+    K = J.shape[0] // 4
+    step_size = 24
+    cols = J.shape[1]
+    
+    # Preallocate arrays for source and target of edges
+    max_possible_edges = (cols // step_size) * K * K  
+    source = np.full(max_possible_edges, -1, dtype=np.int64)
+    target = np.full(max_possible_edges, -1, dtype=np.int64)
+    edge_count = 0
+
+    for idx in nb.prange(0, cols//step_size):
+        col_idx = idx * step_size
+        contacts = np.where(J[:, col_idx: col_idx+step_size])[0]
+        contacts = np.unique(contacts // 4)
+        for i, contact_i in enumerate(contacts):
+            for j, contact_j in enumerate(contacts):
+                if i < j:
+                    source[edge_count] = contact_i
+                    target[edge_count] = contact_j
+                    edge_count += 1
+
+    # Trim the arrays to the actual size
+    valid_indices = source != -1
+    filtered_source = source[valid_indices]
+    filtered_source = target[valid_indices]
+
+    return filtered_source, filtered_source
+
+
+def _greedy_graph_coloring(G):
+    """ Greedy graph coloring algorithm
+
+    Args:
+        G (Graph): The contact grpah is created by on the contact jacobi matrix.
```

Comment:
created "from" the contact Jacobi matrix?

Reply:
No problem.